File: LeetCodeProblemsCMake/Problems/LeetCode229MajorityElementII.cpp
//
// Created by saeed on 6/30/25.
//

#include "LeetCode229MajorityElementII.h"
using namespace std;

vector<int> LeetCode229MajorityElementII::majorityElement(vector<int> &nums) {
    int length = static_cast<int>(nums.size());
    int firstCandidate = 0;
    int secondCandidate = 0;
    int firstVotes{0};
    int secondVotes{0};
    for (auto num: nums) {
        if (num == firstCandidate) {
            firstVotes++;
        } else if (num == secondCandidate) {
            secondVotes++;
        } else if (firstVotes == 0) {
            firstCandidate = num;
            firstVotes = 1;
        } else if (secondVotes == 0) {
            secondCandidate = num;
            secondVotes = 1;
        } else {
            firstVotes--;
            secondVotes--;
        }
    }
    firstVotes = 0;
    secondVotes = 0;
    for (auto num: nums) {
        if (num == firstCandidate) {
            firstVotes++;
        } else if (num == secondCandidate) {
            secondVotes++;
        }
    }
    vector<int> result{};
    if (firstVotes>length/3)
        result.push_back(firstCandidate);
    if (secondVotes>length/3)
        result.push_back(secondCandidate);
    return result;
}

File: LeetCodeProblemsGo/LeetCode229MajorityElementII/LeetCode229MajorityElementII.go
package LeetCode229MajorityElementII

func majorityElement(nums []int) []int {
	lengthThird := len(nums) / 3
	firstCandiate := 0
	firstVotes := 0
	secondCandidate := 0
	secondVotes := 0
	for _, num := range nums {
		if num == firstCandiate {
			firstVotes++
			continue
		}
		if num == secondCandidate {
			secondVotes++
			continue
		}
		if firstVotes == 0 {
			firstCandiate = num
			firstVotes++
			continue
		}
		if secondVotes == 0 {
			secondCandidate = num
			secondVotes++
			continue
		}
		firstVotes--
		secondVotes--
	}
	result := []int{}
	firstVotes = 0
	secondVotes = 0
	for _, num := range nums {
		if num == firstCandiate {
			firstVotes++
			continue
		}
		if num == secondCandidate {
			secondVotes++
			continue
		}
	}
	if firstVotes > lengthThird {
		result = append(result, firstCandiate)
	}
	if secondVotes > lengthThird {
		result = append(result, secondCandidate)
	}
	return result
}
